File: ios/Sources/KlinikPatientsFeature/NewPatientModel.swift
import Foundation
import KlinikAPI
import KlinikCore

public enum NewPatientPhase: Sendable, Equatable {
    case editing
    case saving
    /// Opened, and carrying the number the clinic will write on paper.
    case created(mrn: String, id: String)
    case failed(String)
}

public struct NewPatientState: Sendable, Equatable {
    public var phase: NewPatientPhase = .editing
    public var error: String?

    public init() {}
}

/**
 * Opening a patient file (spec M2).
 *
 * The file number is not asked for and cannot be supplied: the server
 * allocates it. A client that could choose one could collide with a file that
 * already exists, and that number is what the clinic writes on paper and says
 * on the telephone.
 *
 * Validation is done here rather than only by the server, for one reason —
 * a nurse who fills a form, waits, and is then told the country code is wrong
 * has been made to wait for something the screen already knew.
 */
public actor NewPatientModel {
    private let api: PatientsAPI

    private(set) public var state = NewPatientState()

    public init(api: PatientsAPI) {
        self.api = api
    }

    public func currentState() -> NewPatientState { state }

    /// What the form must have before it is worth sending. Nil when it is fine.
    public static func problem(firstName: String, lastName: String, country: String) -> String? {
        if firstName.trimmingCharacters(in: .whitespaces).isEmpty
            || lastName.trimmingCharacters(in: .whitespaces).isEmpty {
            return L10n.string("patient.needName")
        }

        // Two letters, because the server drives language and discharge advice
        // off this and a wrong code sends somebody the wrong emergency number.
        let code = country.trimmingCharacters(in: .whitespaces)
        if code.count != 2 || !code.allSatisfy(\.isLetter) {
            return L10n.string("patient.needCountry")
        }

        return nil
    }

    @discardableResult
    public func create(
        firstName: String,
        lastName: String,
        birthDate: Date,
        sex: String,
        country: String,
        city: String?,
        referralSource: String?
    ) async -> Bool {
        guard state.phase != .saving else { return false }

        if let problem = NewPatientModel.problem(
            firstName: firstName,
            lastName: lastName,
            country: country
        ) {
            state.error = problem
            return false
        }

        state.phase = .saving
        state.error = nil

        do {
            let patient = try await api.create(
                NewPatient(
                    firstName: firstName.trimmingCharacters(in: .whitespaces),
                    lastName: lastName.trimmingCharacters(in: .whitespaces),
                    birthDate: birthDate,
                    sex: sex,
                    country: country.trimmingCharacters(in: .whitespaces).uppercased(),
                    city: city?.isEmpty == true ? nil : city,
                    referralSource: referralSource?.isEmpty == true ? nil : referralSource
                )
            )

            state.phase = .created(mrn: patient.mrn, id: patient.id)
        } catch let error as APIError {
            state.phase = .editing
            state.error = L10n.message(for: error)
            return false
        } catch {
            state.phase = .editing
            state.error = L10n.string("error.server")
            return false
        }

        return true
    }
}

File: android/feature/patients/src/main/kotlin/xyz/klinik/feature/patients/NewPatientModel.kt
package xyz.klinik.feature.patients

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import xyz.klinik.network.ApiError
import xyz.klinik.network.NewPatient
import xyz.klinik.network.Patient
import xyz.klinik.network.PatientsApi
import xyz.klinik.network.UiText
import xyz.klinik.network.uiText

/** What the form holds before it is sent. */
data class NewPatientDraft(
    val firstName: String = "",
    val lastName: String = "",
    val birthDate: String = "",
    val sex: String = "FEMALE",
    val country: String = "",
    val city: String = "",
    val referralSource: String = "",
)

/** Why the form would be refused, before it is sent. */
sealed interface NewPatientProblem {
    data object NoName : NewPatientProblem
    data object BadCountry : NewPatientProblem

    val stringKey: String
        get() = when (this) {
            NoName -> "patient.needName"
            BadCountry -> "patient.needCountry"
        }
}

data class NewPatientState(
    val draft: NewPatientDraft = NewPatientDraft(),
    val busy: Boolean = false,
    /** The file that was opened, so the screen can say which. */
    val created: Patient? = null,
    val error: UiText? = null,
) {
    val problems: List<NewPatientProblem>
        get() = buildList {
            if (draft.firstName.isBlank() || draft.lastName.isBlank()) add(NewPatientProblem.NoName)

            // Two letters, because it is an ISO 3166-1 alpha-2 code and the
            // server refuses anything else — after the form was filled in.
            if (draft.country.trim().length != 2) add(NewPatientProblem.BadCountry)
        }

    val canSubmit: Boolean get() = problems.isEmpty() && !busy && draft.birthDate.isNotBlank()
}

/**
 * Opening a file (spec M2).
 *
 * The five required fields are required by the server, and the form says so
 * while somebody types rather than after they submit. Country is one of them
 * because it decides the language the patient is written to in and which
 * discharge advice they get — a file with no country is one the clinic has to
 * guess about.
 *
 * Nothing clinical is asked for here. A file is opened with an identity; what
 * was operated on and when belongs to the record, not to the moment somebody
 * writes a name down.
 */
class NewPatientModel(private val api: PatientsApi) {
    private val _state = MutableStateFlow(NewPatientState())
    val state: StateFlow<NewPatientState> = _state.asStateFlow()

    fun edit(change: NewPatientDraft.() -> NewPatientDraft) {
        _state.value = _state.value.copy(draft = _state.value.draft.change(), error = null)
    }

    suspend fun create(): Patient? {
        if (!_state.value.canSubmit) return null

        val draft = _state.value.draft

        _state.value = _state.value.copy(busy = true, error = null)

        return try {
            val patient = api.create(
                NewPatient(
                    firstName = draft.firstName.trim(),
                    lastName = draft.lastName.trim(),
                    birthDate = draft.birthDate.trim(),
                    sex = draft.sex,
                    country = draft.country.trim().uppercase(),
                    city = draft.city.trim().ifEmpty { null },
                    referralSource = draft.referralSource.trim().ifEmpty { null },
                ),
            )

            _state.value = _state.value.copy(busy = false, created = patient)
            patient
        } catch (error: Throwable) {
            // The form stays filled in. Retyping a name and a date of birth
            // because the network dropped is how records get entered twice.
            _state.value = _state.value.copy(busy = false, error = messageFor(error))
            null
        }
    }

    private fun messageFor(error: Throwable): UiText =
        (error as? ApiError)?.uiText() ?: UiText.Key("error.server")
}
